frontend/src/scenes/surveys/surveyLogic.tsx

```diff
@@ -177,6 +178,233 @@ function duplicateExistingSurvey(survey: Survey | NewSurvey): Partial<Survey> {
     }
 }
 
+// single question and rating are effectively the same, since we count the frequency of each choice
+export interface ChoiceQuestionProcessedData {
+    type: SurveyQuestionType.SingleChoice | SurveyQuestionType.Rating | SurveyQuestionType.MultipleChoice
+    data?: { label: string; value: number; isPredefined: boolean }[]
+    totalResponses: number
+}
+
+export interface OpenQuestionProcessedData {
+    type: SurveyQuestionType.Open
+    data: { distinctId: string; response: string; personProperties?: Record<string, any> }[]
+    totalResponses: number
+}
+
+export type QuestionProcessedData = ChoiceQuestionProcessedData | OpenQuestionProcessedData
+
+interface ResponsesByQuestion {
+    [questionId: string]: QuestionProcessedData
+}
+
+export interface ConsolidatedSurveyResults {
+    responsesByQuestion: {
+        [questionId: string]: QuestionProcessedData
+    }
+}
+
+function isEmptyOrUndefined(value: any): boolean {
+    return value === null || value === undefined || value === ''
+}
+
+function isQuestionOpenChoice(question: SurveyQuestion, choiceIndex: number): boolean {
+    if (question.type !== SurveyQuestionType.SingleChoice && question.type !== SurveyQuestionType.MultipleChoice) {
+        return false
+    }
+    return !!(choiceIndex === question.choices.length - 1 && question?.hasOpenChoice)
+}
+
+// Extract question processors into separate functions for better maintainability
```

Comment:
the new top-level functions here, plus `consolidatedSurveyResults` in the kea logic are the main ones that require review. this is the new logic for normalizing the data for displaying in the charts